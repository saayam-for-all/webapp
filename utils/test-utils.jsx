import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import authReducer from "#redux/features/authentication/authSlice";

// redux state for logged in user
export const MOCK_STATE_LOGGED_IN = {
  auth: { user: { userId: "mockUser" }, idToken: "mockToken" },
};
// redux state for logged out user
export const MOCK_STATE_LOGGED_OUT = { auth: { user: null } };

export function renderWithProviders(ui, extendedRenderOptions = {}) {
  const {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: {
        auth: authReducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
