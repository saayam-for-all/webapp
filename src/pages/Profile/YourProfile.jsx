import React, { useState, useEffect, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import { updateUserAttributes } from "aws-amplify/auth";
import { useTranslation } from "react-i18next";
import PHONECODESEN from "../../utils/phone-codes-en";
import { getPhoneCodeslist } from "../../utils/utils";
import CountryList from "react-select-country-list";
import { FiPhoneCall, FiVideo } from "react-icons/fi";
import CallModal from "./CallModal.jsx";
import { updateUserProfileSuccess } from "../../redux/features/authentication/authSlice";

function YourProfile({ setHasUnsavedChanges }) {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const [isEditing, setIsEditing] = useState(false);
  const [isCallModalOpen, setIsCallModalOpen] = useState(false);
  const [callType, setCallType] = useState("audio");
  const [loading, setLoading] = useState(false);
  const firstNameRef = useRef(null);
  const countries = CountryList().getData();

  const user = useSelector((state) => state.auth.user);

  const [profileInfo, setProfileInfo] = useState({
    firstName: user?.given_name || "",
    lastName: user?.family_name || "",
    email: user?.email || "",
    phone: user?.phone_number || "",
    phoneCountryCode: "US",
    country: user?.zoneinfo || "",
  });

  useEffect(() => {
    if (user) {
      setProfileInfo({
        firstName: user.given_name || "",
        lastName: user.family_name || "",
        email: user.email || "",
        phone: user.phone_number || "",
        phoneCountryCode: "US",
        country: user.zoneinfo || "",
      });
    }
  }, [user]);

  const handleInputChange = (name, value) => {
    setProfileInfo({ ...profileInfo, [name]: value });
    setHasUnsavedChanges(true);
  };

  const handleSave = async () => {
    try {
      setLoading(true);
      setIsEditing(false);
      setHasUnsavedChanges(false);

      // Prepare updated attributes for Cognito
      const updatedAttributes = {
        given_name: profileInfo.firstName,
        family_name: profileInfo.lastName,
        email: profileInfo.email,
        phone_number: profileInfo.phone,
        "custom:Country": profileInfo.country,
      };

      console.log("Updating Cognito with:", updatedAttributes);
      await updateUserAttributes(updatedAttributes);

      // Update Redux state
      dispatch(updateUserProfileSuccess(updatedAttributes));

      console.log("Profile successfully updated!");
    } catch (error) {
      console.error("Error updating Cognito attributes:", error);
      alert("Failed to save changes. Please try again.");
    } finally {
      setLoading(false);
    }
  };

  return (
    <div className="flex flex-col border p-6 rounded-lg w-full">
      {/* First Name and Last Name */}
      <div className="grid grid-cols-2 gap-4 mb-6">
        <div>
          <label className="block tracking-wide text-gray-700 text-xs font-bold mb-2">
            {t("FIRST_NAME")}
          </label>
          {isEditing ? (
            <input
              ref={firstNameRef}
              type="text"
              value={profileInfo.firstName}
              onChange={(e) => handleInputChange("firstName", e.target.value)}
              className="block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 focus:outline-none"
            />
          ) : (
            <p className="text-lg text-gray-900">{profileInfo.firstName}</p>
          )}
        </div>
        <div>
          <label className="block tracking-wide text-gray-700 text-xs font-bold mb-2">
            {t("LAST_NAME")}
          </label>
          {isEditing ? (
            <input
              type="text"
              value={profileInfo.lastName}
              onChange={(e) => handleInputChange("lastName", e.target.value)}
              className="block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 focus:outline-none"
            />
          ) : (
            <p className="text-lg text-gray-900">{profileInfo.lastName}</p>
          )}
        </div>
      </div>

      {/* Email */}
      <div className="mb-6">
        <label className="block tracking-wide text-gray-700 text-xs font-bold mb-2">
          {t("EMAIL")}
        </label>
        {isEditing ? (
          <input
            type="email"
            value={profileInfo.email}
            onChange={(e) => handleInputChange("email", e.target.value)}
            className="block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 focus:outline-none"
          />
        ) : (
          <p className="text-lg text-gray-900">{profileInfo.email}</p>
        )}
      </div>

      {/* Phone Number */}
      <div className="mb-6">
        <label className="block tracking-wide text-gray-700 text-xs font-bold mb-2">
          {t("PHONE_NUMBER")}
        </label>
        <div className="flex items-center space-x-2">
          {isEditing ? (
            <input
              type="text"
              value={profileInfo.phone}
              onChange={(e) => handleInputChange("phone", e.target.value)}
              className="block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 focus:outline-none"
            />
          ) : (
            <p className="text-lg text-gray-900">
              {profileInfo.phoneCountryCode} {profileInfo.phone}
            </p>
          )}
          <FiPhoneCall
            className="text-gray-500 cursor-pointer hover:text-gray-700"
            onClick={() => setIsCallModalOpen(true)}
          />
          <FiVideo
            className="text-gray-500 cursor-pointer hover:text-gray-700"
            onClick={() => setIsCallModalOpen(true)}
          />
        </div>
      </div>

      {/* Country */}
      <div className="mb-6">
        <label className="block tracking-wide text-gray-700 text-xs font-bold mb-2">
          {t("COUNTRY")}
        </label>
        {isEditing ? (
          <input
            type="text"
            value={profileInfo.country}
            onChange={(e) => handleInputChange("country", e.target.value)}
            className="block w-full bg-gray-200 text-gray-700 border border-gray-200 rounded py-3 px-4 focus:outline-none"
          />
        ) : (
          <p className="text-lg text-gray-900">{profileInfo.country}</p>
        )}
      </div>

      {/* Edit / Save Buttons */}
      <div className="flex justify-center mt-6">
        {!isEditing ? (
          <button
            className="py-2 px-4 bg-blue-500 text-white rounded-md hover:bg-blue-600"
            onClick={() => setIsEditing(true)}
          >
            {t("EDIT")}
          </button>
        ) : (
          <>
            <button
              className="py-2 px-4 bg-blue-500 text-white rounded-md mr-2 hover:bg-blue-600"
              onClick={handleSave}
              disabled={loading}
            >
              {loading ? "Saving..." : t("SAVE")}
            </button>
            <button
              className="py-2 px-4 bg-gray-500 text-white rounded-md hover:bg-gray-600"
              onClick={() => setIsEditing(false)}
            >
              {t("CANCEL")}
            </button>
          </>
        )}
      </div>

      {/* Call Modal */}
      <CallModal
        isOpen={isCallModalOpen}
        onClose={() => setIsCallModalOpen(false)}
        callType={callType}
      />
    </div>
  );
}

export default YourProfile;
