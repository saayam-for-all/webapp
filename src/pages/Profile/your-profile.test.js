import { render } from "@testing-library/react";
import YourProfile from "./YourProfile";
import { MOCK_STATE_LOGGED_IN, renderWithProviders } from "#utils/test-utils";

// TODO: mock CallModal after adding tests for CallModal

describe("YourProfile", () => {
  it("renders correctly", () => {
    const tree = renderWithProviders(<YourProfile />);
    expect(tree).toMatchSnapshot();
  });
});
