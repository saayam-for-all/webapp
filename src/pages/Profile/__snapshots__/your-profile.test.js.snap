// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YourProfile renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col border p-6 rounded-lg w-full"
      >
        <div
          class="grid grid-cols-2 gap-4 mb-6"
        >
          <div>
            <label
              class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
            >
              mockTranslate(FIRST_NAME)
            </label>
            <p
              class="text-lg text-gray-900"
            />
          </div>
          <div>
            <label
              class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
            >
              mockTranslate(LAST_NAME)
            </label>
            <p
              class="text-lg text-gray-900"
            />
          </div>
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
          >
            mockTranslate(EMAIL)
          </label>
          <p
            class="text-lg text-gray-900"
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
          >
            mockTranslate(PHONE_NUMBER)
          </label>
          <div
            class="flex items-center space-x-2"
          >
            <p
              class="text-lg text-gray-900"
            >
              US
               
            </p>
            <svg
              class="text-gray-500 cursor-pointer hover:text-gray-700"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.05 5A5 5 0 0 1 19 8.95M15.05 1A9 9 0 0 1 23 8.94m-1 7.98v3a2 2 0 0 1-2.18 2 19.79 19.79 0 0 1-8.63-3.07 19.5 19.5 0 0 1-6-6 19.79 19.79 0 0 1-3.07-8.67A2 2 0 0 1 4.11 2h3a2 2 0 0 1 2 1.72 12.84 12.84 0 0 0 .7 2.81 2 2 0 0 1-.45 2.11L8.09 9.91a16 16 0 0 0 6 6l1.27-1.27a2 2 0 0 1 2.11-.45 12.84 12.84 0 0 0 2.81.7A2 2 0 0 1 22 16.92z"
              />
            </svg>
            <svg
              class="text-gray-500 cursor-pointer hover:text-gray-700"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="23 7 16 12 23 17 23 7"
              />
              <rect
                height="14"
                rx="2"
                ry="2"
                width="15"
                x="1"
                y="5"
              />
            </svg>
          </div>
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
          >
            mockTranslate(COUNTRY)
          </label>
          <p
            class="text-lg text-gray-900"
          />
        </div>
        <div
          class="flex justify-center mt-6"
        >
          <button
            class="py-2 px-4 bg-blue-500 text-white rounded-md hover:bg-blue-600"
          >
            mockTranslate(EDIT)
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col border p-6 rounded-lg w-full"
    >
      <div
        class="grid grid-cols-2 gap-4 mb-6"
      >
        <div>
          <label
            class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
          >
            mockTranslate(FIRST_NAME)
          </label>
          <p
            class="text-lg text-gray-900"
          />
        </div>
        <div>
          <label
            class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
          >
            mockTranslate(LAST_NAME)
          </label>
          <p
            class="text-lg text-gray-900"
          />
        </div>
      </div>
      <div
        class="mb-6"
      >
        <label
          class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
        >
          mockTranslate(EMAIL)
        </label>
        <p
          class="text-lg text-gray-900"
        />
      </div>
      <div
        class="mb-6"
      >
        <label
          class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
        >
          mockTranslate(PHONE_NUMBER)
        </label>
        <div
          class="flex items-center space-x-2"
        >
          <p
            class="text-lg text-gray-900"
          >
            US
             
          </p>
          <svg
            class="text-gray-500 cursor-pointer hover:text-gray-700"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.05 5A5 5 0 0 1 19 8.95M15.05 1A9 9 0 0 1 23 8.94m-1 7.98v3a2 2 0 0 1-2.18 2 19.79 19.79 0 0 1-8.63-3.07 19.5 19.5 0 0 1-6-6 19.79 19.79 0 0 1-3.07-8.67A2 2 0 0 1 4.11 2h3a2 2 0 0 1 2 1.72 12.84 12.84 0 0 0 .7 2.81 2 2 0 0 1-.45 2.11L8.09 9.91a16 16 0 0 0 6 6l1.27-1.27a2 2 0 0 1 2.11-.45 12.84 12.84 0 0 0 2.81.7A2 2 0 0 1 22 16.92z"
            />
          </svg>
          <svg
            class="text-gray-500 cursor-pointer hover:text-gray-700"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="23 7 16 12 23 17 23 7"
            />
            <rect
              height="14"
              rx="2"
              ry="2"
              width="15"
              x="1"
              y="5"
            />
          </svg>
        </div>
      </div>
      <div
        class="mb-6"
      >
        <label
          class="block tracking-wide text-gray-700 text-xs font-bold mb-2"
        >
          mockTranslate(COUNTRY)
        </label>
        <p
          class="text-lg text-gray-900"
        />
      </div>
      <div
        class="flex justify-center mt-6"
      >
        <button
          class="py-2 px-4 bg-blue-500 text-white rounded-md hover:bg-blue-600"
        >
          mockTranslate(EDIT)
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
